common/src/main/java/com/microsoft/identity/common/internal/result/AdalBrokerResultAdapter.java

```diff
@@ -0,0 +1,343 @@
+    }
+
+    private void setClientExceptionPropertiesToBundle(@NonNull final Bundle resultBundle,
+                                                      @NonNull final ClientException clientException) {
+
+        Logger.verbose(TAG , "Setting properties from ClientException.");
+
+        if (clientException.getErrorCode().equalsIgnoreCase(ErrorStrings.DEVICE_NETWORK_NOT_AVAILABLE)) {
+
+            setErrorToResultBundle(
+                    resultBundle,
+                    AccountManager.ERROR_CODE_NETWORK_ERROR,
+                    ADALError.DEVICE_CONNECTION_IS_NOT_AVAILABLE.getDescription()
+            );
+
+        } else if (clientException.getErrorCode().equalsIgnoreCase(
+                ErrorStrings.NO_NETWORK_CONNECTION_POWER_OPTIMIZATION)) {
+
+            setErrorToResultBundle(
+                    resultBundle,
+                    AccountManager.ERROR_CODE_NETWORK_ERROR,
+                    ADALError.NO_NETWORK_CONNECTION_POWER_OPTIMIZATION.getDescription()
+            );
+
+        }
+    }
+
+
+    /**
+     * Helper method to set Service Exception to Bundle
+     */
+    private void setServiceExceptionPropertiesToBundle(@NonNull final Bundle resultBundle,
+                                                       @NonNull final ServiceException serviceException) {
+
+        Logger.verbose(TAG , "Setting properties from ServiceException.");
+
+        if (null != serviceException.getHttpResponseBody()) {
+            resultBundle.putString(
+                    AuthenticationConstants.OAuth2.HTTP_RESPONSE_BODY,
+                    serviceException.getHttpResponseBody().toString()
+            );
+        }
+
+        if (null != serviceException.getHttpResponseHeaders()) {
+            resultBundle.putString(
+                    AuthenticationConstants.OAuth2.HTTP_RESPONSE_HEADER,
+                    HeaderSerializationUtil.toJson(
+                            serviceException.getHttpResponseHeaders()
+                    )
+            );
+        }
+        resultBundle.putInt(
+                AuthenticationConstants.OAuth2.HTTP_STATUS_CODE,
+                serviceException.getHttpStatusCode()
+        );
+
+        if (serviceException instanceof IntuneAppProtectionPolicyRequiredException) {
+            setIntuneAppProtectionPropertiesToBundle(
+                    resultBundle,
+                    (IntuneAppProtectionPolicyRequiredException) serviceException
+            );
+        }
+
+        if (serviceException.getErrorCode().equalsIgnoreCase(
+                AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT) ||
+                serviceException.getErrorCode().equalsIgnoreCase(
+                        AuthenticationConstants.OAuth2ErrorCode.INTERACTION_REQUIRED)) {
+
+            resultBundle.putString(
+                    AuthenticationConstants.OAuth2.ERROR,
+                    ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED.getDescription()
+            );
```

Comment:
Should it pass the sub error code as well for service exceptions?

Reply:
Adal never reads or looks for sub errors or the old broker never sent one, so these changes are to make sure we send how it was before.

We can add these for Msal

Reply:
It would definitely help the devs for debugging, as it provides more information regarding the error. 

So +1 for MSAL (I understand that we're no longer introducing new things to ADAL, and this is more like a 'good to have' thing).